panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,688 @@
+                                         on_automatic_updates_switch_changed_cb,
+                                         self);
+    }
+}
+
+static void
+load_tariff (CcUpdatesPanel *self)
+{
+  g_autoptr(MwtTariffLoader) tariff_loader = NULL;
+  g_autoptr(GInputStream) input_stream = NULL;
+  g_autoptr(GBytes) bytes = NULL;
+  g_autoptr(GFile) tariff_file = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *tariff_path = NULL;
+
+  g_debug ("Loading tariff");
+
+  tariff_path = g_build_filename (g_get_user_config_dir (), SYSTEM_TARIFF_FILENAME, NULL);
+
+  if (!g_file_test (tariff_path, G_FILE_TEST_EXISTS))
+    {
+      g_debug ("Could not open test tariff file");
+      return;
+    }
+
+  if (!g_file_test (tariff_path, G_FILE_TEST_IS_REGULAR))
+    {
+      g_debug ("Test tariff file is not a regular file");
+      return;
+    }
+
+  tariff_file = g_file_new_for_path (tariff_path);
+  input_stream = G_INPUT_STREAM (g_file_read (tariff_file, NULL, &error));
+  if (error)
+    {
+      g_warning ("Error reading test tariff file: %s", error->message);
+      return;
+    }
+
+  bytes = g_input_stream_read_bytes (input_stream, 8192, NULL, &error);
+  if (error)
+    {
+      g_warning ("Error reading test tariff file: %s", error->message);
+      return;
+    }
+
+  tariff_loader = mwt_tariff_loader_new ();
+  mwt_tariff_loader_load_from_bytes (tariff_loader, bytes, &error);
+  if (error)
+    {
+      g_warning ("Error loading test tariff: %s", error->message);
+      return;
+    }
+
+  g_signal_handlers_block_by_func (self->tariff_editor, on_tariff_changed_cb, self);
+  g_signal_handlers_block_by_func (self->scheduled_updates_switch, on_scheduled_updates_switch_changed_cb, self);
+
+  cc_tariff_editor_load_tariff (self->tariff_editor,
+                                mwt_tariff_loader_get_tariff (tariff_loader),
+                                &error);
+
+  if (error)
+    {
+      g_warning ("Error loading test tariff: %s", error->message);
+    }
+  else
+    {
+      gtk_switch_set_active (GTK_SWITCH (self->scheduled_updates_switch), TRUE);
+    }
+
+
+  g_signal_handlers_unblock_by_func (self->scheduled_updates_switch, on_scheduled_updates_switch_changed_cb, self);
+  g_signal_handlers_unblock_by_func (self->tariff_editor, on_tariff_changed_cb, self);
+
+  g_input_stream_close (input_stream, NULL, &error);
+  if (error)
+    g_warning ("Error closing test tariff file: %s", error->message);
+}
+
+static void
+save_tariff (CcUpdatesPanel *self)
+{
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *tariff_path = NULL;
+  GBytes *tariff_bytes;
+
+  tariff_path = g_build_filename (g_get_user_config_dir (), SYSTEM_TARIFF_FILENAME, NULL);
+  tariff_bytes = cc_tariff_editor_get_tariff_as_bytes (self->tariff_editor);
+
+  if (!tariff_bytes || !gtk_switch_get_active (GTK_SWITCH (self->scheduled_updates_switch)))
+    {
+      /* Remove the tariff file if NULL */
+      if (g_file_test (tariff_path, G_FILE_TEST_EXISTS))
```

Comment:
Again, using `g_file_test()` introduces a TOCTTOU race. You can just check the error value from `g_file_delete()`. (You would have to handle errors from `g_file_delete()` anyway.)